Apply `org.gradle.api.plugins.JvmEcosystemPlugin` in  `eu.xenit.docker-alfresco` plugin

For new alfresco builds, we use `platform()` a lot, because Alfresco has stopped giving WARs sensible versions.

Instead, we have to resort to this to rely on the acs-packaging platform which does have a sensible version:

```groovy
dependencies {
    baseAlfrescoWar platform("org.alfresco:acs-packaging:$alfrescoVersion")
    baseAlfrescoWar 'org.alfresco:content-services@war'
}
```

This is all fine, unless your project does not have the `JvmEcosystemPlugin` applied (or more commonly, the `java` plugin, which applies this `JvmEcosystemPlugin`)

Without this ecosystem plugin applied, you get this very nice and confusing error, which leaves you scratching your head:

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':createDockerFile'.
> Could not resolve all files for configuration ':baseAlfrescoWar'.
   > Could not find org.alfresco:content-services:.
     Required by:
         project :

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 10s
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@
 
 * [#201](https://github.com/xenit-eu/alfresco-docker-gradle-plugin/issues/201) - Make it possible to disable rewriting log4j.properties
 * [#227](https://github.com/xenit-eu/alfresco-docker-gradle-plugin/issues/227) - Improve message for `xyz.amp is not a valid AMP` exception
+* [#231](https://github.com/xenit-eu/alfresco-docker-gradle-plugin/issues/231) - Support for `platform()` & `enforcedPlatform()` with docker-alfresco plugin
 
 ### Fixed
 

--- a/src/integrationTest/examples/alfresco-war-platform/build.gradle
+++ b/src/integrationTest/examples/alfresco-war-platform/build.gradle
@@ -1,0 +1,21 @@
+plugins {
+    id 'eu.xenit.docker-alfresco'
+}
+
+repositories {
+   mavenCentral()
+   maven {
+      url "https://artifacts.alfresco.com/nexus/content/groups/public/"
+   }
+}
+dependencies {
+   baseAlfrescoWar platform('org.alfresco:acs-community-packaging:7.0.0')
+   baseAlfrescoWar "org.alfresco:content-services-community@war"
+}
+
+dockerBuild {
+   alfresco {
+      // Base image used in the FROM of the docker build. Should be a compatible image.
+      baseImage = "tomcat:7-jre8"
+   }
+}

--- a/src/integrationTest/java/eu/xenit/gradle/testrunner/ExampleRunner.java
+++ b/src/integrationTest/java/eu/xenit/gradle/testrunner/ExampleRunner.java
@@ -27,6 +27,11 @@ public class ExampleRunner extends AbstractIntegrationTest {
     }
 
     @Test
+    public void testAlfrescoWarPlatform() throws IOException {
+        testProjectFolder(EXAMPLES.resolve("alfresco-war-platform"));
+    }
+
+    @Test
     public void testAlfrescoWarOnlyLegacy() throws IOException {
         testProjectFolder(EXAMPLES.resolve("alfresco-war-only-legacy"));
     }

--- a/src/main/java/eu/xenit/gradle/docker/alfresco/DockerAlfrescoPlugin.java
+++ b/src/main/java/eu/xenit/gradle/docker/alfresco/DockerAlfrescoPlugin.java
@@ -1,6 +1,7 @@
 package eu.xenit.gradle.docker.alfresco;
 
 import com.bmuschko.gradle.docker.tasks.image.Dockerfile;
+import eu.xenit.gradle.docker.alfresco.internal.gradle.backports.JvmEcosystemPluginBackport;
 import eu.xenit.gradle.docker.alfresco.tasks.InjectFilesInWarTask;
 import eu.xenit.gradle.docker.alfresco.tasks.InstallAmpsInWarTask;
 import eu.xenit.gradle.docker.alfresco.tasks.MergeWarsTask;
@@ -14,6 +15,7 @@ import eu.xenit.gradle.docker.alfresco.tasks.extension.internal.DockerfileWithLa
 import eu.xenit.gradle.docker.alfresco.tasks.extension.internal.DockerfileWithWarsExtensionImpl;
 import eu.xenit.gradle.docker.core.DockerExtension;
 import eu.xenit.gradle.docker.core.DockerPlugin;
+import eu.xenit.gradle.docker.internal.GradleVersionRequirement;
 import java.util.ArrayList;
 import java.util.List;
 import org.alfresco.repo.module.tool.WarHelperImpl;
@@ -21,6 +23,7 @@ import org.gradle.api.NamedDomainObjectProvider;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.plugins.JvmEcosystemPlugin;
 import org.gradle.api.provider.Provider;
 import org.gradle.api.tasks.TaskProvider;
 
@@ -50,6 +53,11 @@ public class DockerAlfrescoPlugin implements Plugin<Project> {
         createConfiguration(project, SHARE_SM);
 
         project.getPluginManager().apply(DockerPlugin.class);
+        if(GradleVersionRequirement.isAtLeast("6.7.0", "support platform() & enforcedPlatform() in dependencies without hacks")) {
+            project.getPluginManager().apply(JvmEcosystemPlugin.class);
+        } else {
+            project.getPluginManager().apply(JvmEcosystemPluginBackport.class);
+        }
 
         DockerExtension dockerExtension = project.getExtensions().getByType(DockerExtension.class);
         AlfrescoDockerExtension alfrescoDockerExtension = dockerExtension.getExtensions()

--- a/src/main/java/eu/xenit/gradle/docker/alfresco/internal/gradle/backports/JvmEcosystemPluginBackport.java
+++ b/src/main/java/eu/xenit/gradle/docker/alfresco/internal/gradle/backports/JvmEcosystemPluginBackport.java
@@ -1,0 +1,39 @@
+package eu.xenit.gradle.docker.alfresco.internal.gradle.backports;
+
+import eu.xenit.gradle.docker.internal.GradleVersionRequirement;
+import java.lang.reflect.InvocationTargetException;
+import org.gradle.api.Plugin;
+import org.gradle.api.internal.artifacts.dsl.ComponentMetadataHandlerInternal;
+import org.gradle.api.internal.project.ProjectInternal;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.internal.component.external.model.JavaEcosystemVariantDerivationStrategy;
+
+public class JvmEcosystemPluginBackport implements Plugin<ProjectInternal> {
+    private static final Logger LOGGER = Logging.getLogger(JvmEcosystemPluginBackport.class);
+
+    @Override
+    public void apply(ProjectInternal projectInternal) {
+        // Copied and adapted from JavaBasePlugin
+        // https://github.com/gradle/gradle/blob/v6.6.0/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java#L131-L134
+        // https://github.com/gradle/gradle/blob/v6.4.0/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java#L130-L133
+        ComponentMetadataHandlerInternal metadataHandler = (ComponentMetadataHandlerInternal) projectInternal.getDependencies().getComponents();
+        JavaEcosystemVariantDerivationStrategy ecosystemVariantDerivationStrategy = GradleVersionRequirement.ifAtLeast("6.5.0", JvmEcosystemPluginBackport::createWithInstance, JvmEcosystemPluginBackport::createWithReflection);
+        metadataHandler.setVariantDerivationStrategy(ecosystemVariantDerivationStrategy);
+    }
+
+    private static JavaEcosystemVariantDerivationStrategy createWithInstance() {
+        return JavaEcosystemVariantDerivationStrategy.getInstance();
+    }
+
+    private static JavaEcosystemVariantDerivationStrategy createWithReflection() {
+        try {
+            return JavaEcosystemVariantDerivationStrategy.class.getConstructor().newInstance();
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            LOGGER.error("Failed to create JavaEcosystemVariantDerivationStrategy with reflection.", e);
+            GradleVersionRequirement.atLeast("6.5.0", "use platform() & enforcedPlatform() for dependencies");
+        }
+        // unreachable
+        return null;
+    }
+}

--- a/src/main/java/eu/xenit/gradle/docker/internal/GradleVersionRequirement.java
+++ b/src/main/java/eu/xenit/gradle/docker/internal/GradleVersionRequirement.java
@@ -1,34 +1,66 @@
 package eu.xenit.gradle.docker.internal;
 
 import java.util.function.Supplier;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
 import org.gradle.util.GradleVersion;
 
 public final class GradleVersionRequirement {
+    private static final Logger LOGGER  = Logging.getLogger(GradleVersionRequirement.class);
 
     private GradleVersionRequirement() {
     }
 
     static class UnsupportedGradleVersion extends UnsupportedOperationException {
+        private static String formatRequirement(GradleVersion requiredVersion, GradleVersion currentVersion, String feature) {
+            return requiredVersion.toString() + " or later is required" + (feature == null ? ""
+                    : " to " + feature) + ". You are currently using " + currentVersion.toString();
+        }
 
         private UnsupportedGradleVersion(GradleVersion requiredVersion, GradleVersion currentVersion, String feature) {
-            super(requiredVersion.toString() + " or later is required" + (feature == null ? ""
-                    : " to " + feature) + ". You are currently using " + currentVersion.toString());
+            super(formatRequirement(requiredVersion, currentVersion, feature));
         }
     }
 
-    public static void atLeast(GradleVersion gradleVersion, String feature) {
+    public static <T> T ifAtLeast(String gradleVersion, Supplier<T> ifVersion, Supplier<T> ifNotVersion) {
+        return ifAtLeast(GradleVersion.version(gradleVersion), ifVersion, ifNotVersion);
+    }
+
+    public static <T> T ifAtLeast(GradleVersion gradleVersion, Supplier<T> ifVersion, Supplier<T> ifNotVersion) {
+        if(isAtLeast(gradleVersion, null, true)) {
+            return ifVersion.get();
+        } else {
+            return ifNotVersion.get();
+        }
+
+    }
+
+    public static boolean isAtLeast(GradleVersion gradleVersion, String feature) {
+        return isAtLeast(gradleVersion, feature, false);
+    }
+
+    public static boolean isAtLeast(GradleVersion gradleVersion, String feature, boolean silent) {
         if (GradleVersion.current().compareTo(gradleVersion) < 0) {
+            if(!silent) {
+                LOGGER.warn("{}", UnsupportedGradleVersion.formatRequirement(gradleVersion, GradleVersion.current(), feature));
+            }
+            return false;
+        }
+        return true;
+    }
+
+    public static boolean isAtLeast(String gradleVersion, String feature) {
+        return isAtLeast(GradleVersion.version(gradleVersion), feature);
+    }
+
+    public static void atLeast(GradleVersion gradleVersion, String feature) {
+        if (!isAtLeast(gradleVersion, feature, true)) {
             throw new UnsupportedGradleVersion(gradleVersion, GradleVersion.current(), feature);
         }
     }
 
     public static void atLeast(String gradleVersion, String feature) {
         atLeast(GradleVersion.version(gradleVersion), feature);
-    }
-
-    public static <T> T atLeast(GradleVersion gradleVersion, String feature, Supplier<T> callable) {
-        atLeast(gradleVersion, feature);
-        return callable.get();
     }
 
     public static <T> T atLeast(String gradleVersion, String feature, Supplier<T> callable) {

